distsql: panic 'flow already registered'

Ran into the panic below the first time I tried to run a distsql query on a node. Did not run into it again after restart. Nothing useful in the logs, except a long spam of messages like the one below (prior to the query being run).
```
I170112 14:11:20.066619 5270719 vendor/google.golang.org/grpc/server.go:723  grpc: Server.processUnaryRPC failed to write status: stream error: code = 4 desc = "context deadline exceeded"
```

I looked through the code and the only explanation I can come up with is if the distsql planner's nodeID was not set correctly. Then we may end up sending a setup flow request to ourselves AND setup a sync flow locally with the same id.

Investigate the nodeID issue and make this into an error not a panic.


```
panic: flow already registered [recovered]
	panic: SELECT COUNT(*) FROM comments;: flow already registered [recovered]
	panic: SELECT COUNT(*) FROM comments;: flow already registered

goroutine 5064283 [running]:
panic(0x1767a00, 0xc4f5d471e0)
	/home/radu/goroot/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4200b0870)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x6e
panic(0x1767a00, 0xc4f5d471e0)
	/home/radu/goroot/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc4abcf31b7, 0x1e)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:455 +0x149
panic(0x1722600, 0xc4f5d471c0)
	/home/radu/goroot/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*flowRegistry).RegisterFlow(0xc4202285d0, 0xe94ba1002c773502, 0x5fd584bb82d464b0, 0xc4f5a11680, 0xc4c729cc30)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_registry.go:130 +0x328
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start(0xc4f5a11680, 0x1af3fe0)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:350 +0x1f7
github.com/cockroachdb/cockroach/pkg/sql.(*distSQLPlanner).PlanAndRun(0xc4203e6310, 0x7fd97b5218e8, 0xc48798a000, 0xc4ea2e4000, 0x249cfc0, 0xc4af580300, 0xc46fe48000, 0x0, 0xc4d6dc90c0)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1474 +0x1366
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execDistSQL(0xc4201ba9a0, 0xc43c922d70, 0x249cfc0, 0xc4af580300, 0xc50af37ee8, 0x0, 0x0)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1222 +0x113
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc4201ba9a0, 0x2494da0, 0xc4dca0bfb0, 0xc43c922d70, 0xc50af30001, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1343 +0x49d
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc4201ba9a0, 0x2494da0, 0xc4dca0bfb0, 0xc43c922d70, 0x101, 0xc43c922c68, 0x0, 0x0, 0x0, 0x0, ...)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1113 +0x2ab
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtsInCurrentTxn(0xc4201ba9a0, 0xc4ed3f0c50, 0x1, 0x1, 0xc43c922d70, 0xc43c922c68, 0x101, 0x0, 0x6, 0xaf38b48, ...)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:850 +0x7db
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc4201ba9a0, 0xc43c922d70, 0x1, 0xc43c922c68, 0xc4ed3f0c90, 0xc4ed3f0c50, 0x1, 0x1, 0x0, 0x0, ...)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:768 +0xf8
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest.func2(0xc4ea2e4000, 0xc4ed3f0c90, 0x0, 0x0)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:621 +0x2b7
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc4ea2e4000, 0xc4ad550101, 0xc420018140, 0xc4ad555880, 0x0, 0x0)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:520 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc4201ba9a0, 0xc43c922c00, 0xc4abcf31b7, 0x1e, 0x0, 0x0, 0x0, 0x0, 0xc4e6bad300)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:643 +0x5b0
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc4201ba9a0, 0xc43c922c00, 0xc4abcf31b7, 0x1e, 0x0, 0x0, 0x0, 0x0, 0xc4abcf3100)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:461 +0x11f
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc49b401800, 0x7fd97b5218e8, 0xc4b0e53950, 0xc4abcf31b7, 0x1e, 0x0, 0x0, 0x0, 0x0, 0x701601, ...)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:792 +0x94
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc49b401800, 0x7fd97b5218e8, 0xc4b0e53950, 0xc49b401828, 0x23, 0x0)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:470 +0xb3
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc49b401800, 0x7fd97b5218e8, 0xc4b0e53950, 0x5400, 0xc420250138, 0x7fd97b5218e8, 0xc4b0e53920, 0x0, 0x0)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:397 +0xa62
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc420250000, 0x7fd97b5218e8, 0xc4b0e53920, 0x249c280, 0xc4d66a0000, 0x0, 0x0)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:309 +0x999
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func8.1(0x249c280, 0xc4d66a0000)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:522 +0x106
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4200b0870, 0xc420042030, 0x249c280, 0xc4d66a0000, 0xc4209c5780)
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
	/home/radu/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x277
```

--- a/pkg/gossip/gossip.go
+++ b/pkg/gossip/gossip.go
@@ -242,6 +242,13 @@ type Gossip struct {
 // The higher level manages the NodeIDContainer instance (which can be shared by
 // various server components). The ambient context is expected to already
 // contain the node ID.
+//
+// grpcServer: The server on which the new Gossip instance will register its RPC
+//   service. Can be nil, in which case the Gossip will not register the
+//   service.
+// rpcContext: The context used to connect to other nodes. Can be nil for tests
+//   that also specify a nil grpcServer and that plan on using the Gossip in a
+//   restricted way by populating it with data manually.
 func New(
 	ambient log.AmbientContext,
 	nodeID *base.NodeIDContainer,
@@ -281,12 +288,21 @@ func New(
 	g.mu.is.registerCallback(MakePrefixPattern(KeyStorePrefix), g.updateStoreMap)
 	g.mu.Unlock()
 
-	RegisterGossipServer(grpcServer, g.server)
+	if grpcServer != nil {
+		RegisterGossipServer(grpcServer, g.server)
+	}
 	return g
 }
 
 // NewTest is a simplified wrapper around New that creates the NodeIDContainer
 // internally. Used for testing.
+//
+// grpcServer: The server on which the new Gossip instance will register its RPC
+//   service. Can be nil, in which case the Gossip will not register the
+//   service.
+// rpcContext: The context used to connect to other nodes. Can be nil for tests
+//   that also specify a nil grpcServer and that plan on using the Gossip in a
+//   restricted way by populating it with data manually.
 func NewTest(
 	nodeID roachpb.NodeID,
 	rpcContext *rpc.Context,

--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -531,20 +531,36 @@ func (dsp *distSQLPlanner) partitionSpans(
 				addr, inAddrMap := planCtx.nodeAddresses[nodeID]
 				if !inAddrMap {
 					addr = replInfo.NodeDesc.Address.String()
-					var err error
-					if dsp.testingKnobs.OverrideHealthCheck != nil {
-						err = dsp.testingKnobs.OverrideHealthCheck(replInfo.NodeDesc.NodeID, addr)
-					} else {
-						err = dsp.rpcContext.ConnHealth(addr)
+					checkNodeHealth := func() error {
+						// Check if the node is still in gossip - i.e. if it hasn't been
+						// decommissioned or overridden by another node at the same address.
+						if _, err := dsp.gossip.GetNodeIDAddress(nodeID); err != nil {
+							log.VEventf(ctx, 1, "not using n%d because gossip doesn't know about it. "+
+								"It might have gone away from the cluster. Gossip said: %s.", nodeID, err)
+							return err
+						}
+
+						var err error
+						if dsp.testingKnobs.OverrideHealthCheck != nil {
+							err = dsp.testingKnobs.OverrideHealthCheck(replInfo.NodeDesc.NodeID, addr)
+						} else {
+							err = dsp.rpcContext.ConnHealth(addr)
+						}
+						if err != nil && err != rpc.ErrNotConnected && err != rpc.ErrNotHeartbeated {
+							// This host is known to be unhealthy. Don't use it (use the gateway
+							// instead). Note: this can never happen for our nodeID (which
+							// always has its address in the nodeMap).
+							log.VEventf(ctx, 1, "marking n%d as unhealthy for this plan: %v", nodeID, err)
+							return err
+						}
+						return nil
 					}
-					if err != nil && err != rpc.ErrNotConnected && err != rpc.ErrNotHeartbeated {
-						// This host is known to be unhealthy. Don't use it (use the gateway
-						// instead). Note: this can never happen for our nodeID (which
-						// always has its address in the nodeMap).
+					if err := checkNodeHealth(); err != nil {
 						addr = ""
-						log.VEventf(ctx, 1, "marking node %d as unhealthy for this plan: %v", nodeID, err)
 					}
-					planCtx.nodeAddresses[nodeID] = addr
+					if err == nil && addr != "" {
+						planCtx.nodeAddresses[nodeID] = addr
+					}
 				}
 				compat := true
 				if addr != "" {
@@ -603,15 +619,8 @@ func (dsp *distSQLPlanner) nodeVersionIsCompatible(
 	nodeID roachpb.NodeID, planVer distsqlrun.DistSQLVersion,
 ) bool {
 	var v distsqlrun.DistSQLVersionGossipInfo
-	if hook := dsp.testingKnobs.OverrideDistSQLVersionCheck; hook != nil {
-		var err error
-		if v, err = hook(nodeID); err != nil {
-			return false
-		}
-	} else {
-		if err := dsp.gossip.GetInfoProto(gossip.MakeDistSQLNodeVersionKey(nodeID), &v); err != nil {
-			return false
-		}
+	if err := dsp.gossip.GetInfoProto(gossip.MakeDistSQLNodeVersionKey(nodeID), &v); err != nil {
+		return false
 	}
 	return distsqlrun.FlowVerIsCompatible(dsp.planVersion, v.MinAcceptedVersion, v.Version)
 }

--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -546,7 +546,7 @@ func (dsp *distSQLPlanner) partitionSpans(
 					}
 					planCtx.nodeAddresses[nodeID] = addr
 				}
-				var compat bool
+				compat := true
 				if addr != "" {
 					// Check if the node's DistSQL version is compatible with this plan.
 					// If it isn't, we'll use the gateway.

--- a/pkg/sql/distsql_physical_planner_test.go
+++ b/pkg/sql/distsql_physical_planner_test.go
@@ -45,6 +45,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/metric"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 )
@@ -718,21 +719,45 @@ func TestPartitionSpans(t *testing.T) {
 		},
 	}
 
+	// We need a mock Gossip to contain addresses for the nodes. Otherwise the
+	// distSQLPlanner will not plan flows on them.
+	testStopper := stop.NewStopper()
+	defer testStopper.Stop(context.TODO())
+	mockGossip := gossip.NewTest(roachpb.NodeID(1), nil /* rpcContext */, nil, /* grpcServer */
+		testStopper, metric.NewRegistry())
+	var nodeDescs []*roachpb.NodeDescriptor
+	for i := 1; i <= 10; i++ {
+		nodeID := roachpb.NodeID(i)
+		desc := &roachpb.NodeDescriptor{
+			NodeID:  nodeID,
+			Address: util.UnresolvedAddr{AddressField: fmt.Sprintf("addr%d", i)},
+		}
+		if err := mockGossip.SetNodeDescriptor(desc); err != nil {
+			t.Fatal(err)
+		}
+		if err := mockGossip.AddInfoProto(
+			gossip.MakeDistSQLNodeVersionKey(nodeID),
+			&distsqlrun.DistSQLVersionGossipInfo{
+				MinAcceptedVersion: distsqlrun.MinAcceptedVersion,
+				Version:            distsqlrun.Version,
+			},
+			0, // ttl - no expiration
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		nodeDescs = append(nodeDescs, desc)
+	}
+
 	for testIdx, tc := range testCases {
 		t.Run(strconv.Itoa(testIdx), func(t *testing.T) {
 			stopper := stop.NewStopper()
 			defer stopper.Stop(context.TODO())
 
-			tsp := &testSpanResolver{}
-			for i := 1; i <= 10; i++ {
-				tsp.nodes = append(tsp.nodes, &roachpb.NodeDescriptor{
-					NodeID: roachpb.NodeID(i),
-					Address: util.UnresolvedAddr{
-						AddressField: fmt.Sprintf("addr%d", i),
-					},
-				})
+			tsp := &testSpanResolver{
+				nodes:  nodeDescs,
+				ranges: tc.ranges,
 			}
-			tsp.ranges = tc.ranges
 
 			dsp := distSQLPlanner{
 				planVersion:  distsqlrun.Version,
@@ -740,6 +765,7 @@ func TestPartitionSpans(t *testing.T) {
 				nodeDesc:     *tsp.nodes[tc.gatewayNode-1],
 				stopper:      stopper,
 				spanResolver: tsp,
+				gossip:       mockGossip,
 				testingKnobs: DistSQLPlannerTestingKnobs{
 					OverrideHealthCheck: func(node roachpb.NodeID, addr string) error {
 						for _, n := range tc.deadNodes {
@@ -748,15 +774,6 @@ func TestPartitionSpans(t *testing.T) {
 							}
 						}
 						return nil
-					},
-					OverrideDistSQLVersionCheck: func(
-						node roachpb.NodeID) (distsqlrun.DistSQLVersionGossipInfo, error,
-					) {
-						// Accept any version.
-						return distsqlrun.DistSQLVersionGossipInfo{
-							MinAcceptedVersion: 0,
-							Version:            1000000,
-						}, nil
 					},
 				},
 			}
@@ -814,9 +831,10 @@ func TestPartitionSpansSkipsIncompatibleNodes(t *testing.T) {
 		// The versions accepted by each node.
 		nodeVersions map[roachpb.NodeID]distsqlrun.DistSQLVersionGossipInfo
 
-		// nodesNotInGossip is the set of nodes for which we're going to pretend
-		// that gossip returns an error when queried about the DistSQL version.
-		nodesNotInGossip map[roachpb.NodeID]struct{}
+		// nodesNotAdvertisingDistSQLVersion is the set of nodes for which gossip is
+		// not going to have information about the supported DistSQL version. This
+		// is to simulate CRDB 1.0 nodes which don't advertise this information.
+		nodesNotAdvertisingDistSQLVersion map[roachpb.NodeID]struct{}
 
 		// expected result: a map of node to list of spans.
 		partitions map[roachpb.NodeID][][2]string
@@ -871,7 +889,7 @@ func TestPartitionSpansSkipsIncompatibleNodes(t *testing.T) {
 					Version:            3,
 				},
 			},
-			nodesNotInGossip: map[roachpb.NodeID]struct{}{
+			nodesNotAdvertisingDistSQLVersion: map[roachpb.NodeID]struct{}{
 				1: {},
 			},
 			partitions: map[roachpb.NodeID][][2]string{
@@ -879,22 +897,47 @@ func TestPartitionSpansSkipsIncompatibleNodes(t *testing.T) {
 			},
 		},
 	}
+
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
 
 			stopper := stop.NewStopper()
 			defer stopper.Stop(context.TODO())
 
-			tsp := &testSpanResolver{}
+			// We need a mock Gossip to contain addresses for the nodes. Otherwise the
+			// distSQLPlanner will not plan flows on them. This Gossip will also
+			// reflect tc.nodesNotAdvertisingDistSQLVersion.
+			testStopper := stop.NewStopper()
+			defer testStopper.Stop(context.TODO())
+			mockGossip := gossip.NewTest(roachpb.NodeID(1), nil /* rpcContext */, nil, /* grpcServer */
+				testStopper, metric.NewRegistry())
+			var nodeDescs []*roachpb.NodeDescriptor
 			for i := 1; i <= 2; i++ {
-				tsp.nodes = append(tsp.nodes, &roachpb.NodeDescriptor{
-					NodeID: roachpb.NodeID(i),
-					Address: util.UnresolvedAddr{
-						AddressField: fmt.Sprintf("addr%d", i),
-					},
-				})
+				nodeID := roachpb.NodeID(i)
+				desc := &roachpb.NodeDescriptor{
+					NodeID:  nodeID,
+					Address: util.UnresolvedAddr{AddressField: fmt.Sprintf("addr%d", i)},
+				}
+				if err := mockGossip.SetNodeDescriptor(desc); err != nil {
+					t.Fatal(err)
+				}
+				if _, ok := tc.nodesNotAdvertisingDistSQLVersion[nodeID]; !ok {
+					verInfo := tc.nodeVersions[nodeID]
+					if err := mockGossip.AddInfoProto(
+						gossip.MakeDistSQLNodeVersionKey(nodeID),
+						&verInfo,
+						0, // ttl - no expiration
+					); err != nil {
+						t.Fatal(err)
+					}
+				}
+
+				nodeDescs = append(nodeDescs, desc)
 			}
-			tsp.ranges = ranges
+			tsp := &testSpanResolver{
+				nodes:  nodeDescs,
+				ranges: ranges,
+			}
 
 			dsp := distSQLPlanner{
 				planVersion:  tc.planVersion,
@@ -902,19 +945,11 @@ func TestPartitionSpansSkipsIncompatibleNodes(t *testing.T) {
 				nodeDesc:     *tsp.nodes[gatewayNode-1],
 				stopper:      stopper,
 				spanResolver: tsp,
+				gossip:       mockGossip,
 				testingKnobs: DistSQLPlannerTestingKnobs{
 					OverrideHealthCheck: func(node roachpb.NodeID, addr string) error {
 						// All the nodes are healthy.
 						return nil
-					},
-					OverrideDistSQLVersionCheck: func(
-						nodeID roachpb.NodeID) (distsqlrun.DistSQLVersionGossipInfo, error,
-					) {
-						if _, ok := tc.nodesNotInGossip[nodeID]; ok {
-							return distsqlrun.DistSQLVersionGossipInfo{},
-								gossip.NewKeyNotPresentError(gossip.MakeDistSQLNodeVersionKey(nodeID))
-						}
-						return tc.nodeVersions[nodeID], nil
 					},
 				},
 			}
@@ -941,5 +976,98 @@ func TestPartitionSpansSkipsIncompatibleNodes(t *testing.T) {
 				t.Errorf("expected partitions:\n  %v\ngot:\n  %v", tc.partitions, resMap)
 			}
 		})
+	}
+}
+
+// Test that a node whose descriptor info is not accessible through gossip is
+// not used. This is to simulate nodes that have been decomisioned and also
+// nodes that have been "replaced" by another node at the same address (which, I
+// guess, is also a type of decomissioning).
+func TestPartitionSpansSkipsNodesNotInGossip(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// The spans that we're going to plan for.
+	span := roachpb.Span{Key: roachpb.Key("A"), EndKey: roachpb.Key("Z")}
+	gatewayNode := roachpb.NodeID(2)
+	ranges := []testSpanResolverRange{{"A", 1}, {"B", 2}, {"C", 1}}
+
+	stopper := stop.NewStopper()
+	defer stopper.Stop(context.TODO())
+
+	mockGossip := gossip.NewTest(roachpb.NodeID(1), nil /* rpcContext */, nil, /* grpcServer */
+		stopper, metric.NewRegistry())
+	var nodeDescs []*roachpb.NodeDescriptor
+	for i := 1; i <= 2; i++ {
+		nodeID := roachpb.NodeID(i)
+		desc := &roachpb.NodeDescriptor{
+			NodeID:  nodeID,
+			Address: util.UnresolvedAddr{AddressField: fmt.Sprintf("addr%d", i)},
+		}
+		if i == 2 {
+			if err := mockGossip.SetNodeDescriptor(desc); err != nil {
+				t.Fatal(err)
+			}
+		}
+		// All the nodes advertise their DistSQL versions. This is to simulate the
+		// "node overridden by another node at the same address" case mentioned in
+		// the test comment - for such a node, the descriptor would be taken out of
+		// the gossip data, but other datums it advertised are left in place.
+		if err := mockGossip.AddInfoProto(
+			gossip.MakeDistSQLNodeVersionKey(nodeID),
+			&distsqlrun.DistSQLVersionGossipInfo{
+				MinAcceptedVersion: distsqlrun.MinAcceptedVersion,
+				Version:            distsqlrun.Version,
+			},
+			0, // ttl - no expiration
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		nodeDescs = append(nodeDescs, desc)
+	}
+	tsp := &testSpanResolver{
+		nodes:  nodeDescs,
+		ranges: ranges,
+	}
+
+	dsp := distSQLPlanner{
+		planVersion:  distsqlrun.Version,
+		st:           cluster.MakeTestingClusterSettings(),
+		nodeDesc:     *tsp.nodes[gatewayNode-1],
+		stopper:      stopper,
+		spanResolver: tsp,
+		gossip:       mockGossip,
+		testingKnobs: DistSQLPlannerTestingKnobs{
+			OverrideHealthCheck: func(node roachpb.NodeID, addr string) error {
+				// All the nodes are healthy.
+				return nil
+			},
+		},
+	}
+
+	planCtx := dsp.NewPlanningCtx(context.Background(), nil /* txn */)
+	partitions, err := dsp.partitionSpans(&planCtx, roachpb.Spans{span})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	resMap := make(map[roachpb.NodeID][][2]string)
+	for _, p := range partitions {
+		if _, ok := resMap[p.node]; ok {
+			t.Fatalf("node %d shows up in multiple partitions", p)
+		}
+		var spans [][2]string
+		for _, s := range p.spans {
+			spans = append(spans, [2]string{string(s.Key), string(s.EndKey)})
+		}
+		resMap[p.node] = spans
+	}
+
+	expectedPartitions :=
+		map[roachpb.NodeID][][2]string{
+			2: {{"A", "Z"}},
+		}
+	if !reflect.DeepEqual(resMap, expectedPartitions) {
+		t.Errorf("expected partitions:\n  %v\ngot:\n  %v", expectedPartitions, resMap)
 	}
 }

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -117,7 +117,7 @@ func (dsp *distSQLPlanner) Run(
 		return err
 	}
 
-	flows := plan.GenerateFlowSpecs(dsp.nodeDesc.NodeID)
+	flows := plan.GenerateFlowSpecs(dsp.nodeDesc.NodeID /* gateway */)
 
 	if logPlanDiagram {
 		log.VEvent(ctx, 1, "creating plan diagram")

--- a/pkg/sql/distsqlplan/span_resolver.go
+++ b/pkg/sql/distsqlplan/span_resolver.go
@@ -320,6 +320,7 @@ func (it *spanResolverIterator) ReplicaInfo(ctx context.Context) (kv.ReplicaInfo
 				// Ignore the error; ask the oracle to pick another replica below.
 				log.VEventf(ctx, 2, "failed to resolve node %d: %s", nodeID, err)
 			} else {
+				repl.ReplicaDescriptor.NodeID = nodeID
 				repl.NodeDesc = nd
 				resolvedLH = true
 			}

--- a/pkg/sql/distsqlrun/flow_registry.go
+++ b/pkg/sql/distsqlrun/flow_registry.go
@@ -22,7 +22,6 @@ import (
 	"golang.org/x/net/context"
 
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
-	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -164,11 +163,10 @@ func (fr *flowRegistry) RegisterFlow(
 	}()
 	entry := fr.getEntryLocked(id)
 	if entry.flow != nil {
-		return util.UnexpectedWithIssueErrorf(
-			12876,
+		return errors.Errorf(
 			"flow already registered: current node ID: %d flowID: %d.\n"+
-				"Current flow:%+v\nExisting flow:%+v",
-			fr.nodeID, f.spec, entry.flow.spec)
+				"Current flow: %+v\nExisting flow: %+v",
+			fr.nodeID, f.spec.FlowID, f.spec, entry.flow.spec)
 	}
 	// Take a reference that will be removed by UnregisterFlow.
 	entry.refCount++

--- a/pkg/sql/executor.go
+++ b/pkg/sql/executor.go
@@ -365,13 +365,6 @@ type DistSQLPlannerTestingKnobs struct {
 	// If OverrideSQLHealthCheck is set, we use this callback to get the health of
 	// a node.
 	OverrideHealthCheck func(node roachpb.NodeID, addrString string) error
-	// If OverrideDistSQLVersionCheck is set, the distSQLPlanner uses this instead
-	// of gossip for figuring out a node's DistSQL version. The callback can
-	// return an error to simulate gossip not having any info for the node. If the
-	// test wants to simulate the node accepting any version, the callback can
-	// return a 0..+inf acceptable version range.
-	OverrideDistSQLVersionCheck func(
-		node roachpb.NodeID) (distsqlrun.DistSQLVersionGossipInfo, error)
 }
 
 // NewExecutor creates an Executor and registers a callback on the

